finesse/hardware/data_file_writer.py

```diff
@@ -0,0 +1,117 @@
+from .temperature import get_hot_bb_temperature_controller_instance
+
+
+def _on_error_occurred(error: BaseException) -> None:
+    """Close file in case of error."""
+    pub.sendMessage("data_file.close")
+
+
+def _get_platform_info() -> dict[str, str]:
+    return {
+        key: getattr(platform, key)() for key in ("platform", "python_version", "node")
+    }
+
+
+def _get_metadata(filename: str) -> dict[str, Any]:
+    return {
+        "encoding": "utf-8",
+        "name": filename,
+        "datetime": datetime.now().astimezone().isoformat(),  # include timezone
+        "system": {
+            "app": {
+                "name": config.APP_NAME,
+                "author": config.APP_AUTHOR,
+                "version": config.APP_VERSION,
+            },
+            "platform": _get_platform_info(),
+        },
+    }
+
+
+class DataFileWriter:
+    """A class for writing sensor data to a CSV file.
+
+    This class is a singleton. Every time a new file needs to be written this instance
+    is reused.
+    """
+
+    def __init__(self) -> None:
+        """Create a new DataFileWriter."""
+        self._writer: Writer
+        """The CSV writer."""
+
+        # Listen to open/close messages
+        pub.subscribe(self.open, "data_file.open")
+        pub.subscribe(self.close, "data_file.close")
+
+        # Listen for error messages
+        pub.subscribe(_on_error_occurred, "data_file.error")
+
+    @pubsub_errors("data_file.error")
+    def open(self, path: Path) -> None:
+        """Open a file at the specified path for writing.
+
+        Args:
+            path: The path of the file to write to
+        """
+        self._writer = Writer(path, _get_metadata(path.name))
+
+        # Write column headers
+        self._writer.writerow(
+            (
+                "Date",
+                "Time",
+                *(f"Temp{i+1}" for i in range(config.NUM_TEMPERATURE_MONITOR_CHANNELS)),
+                "TimeAsSeconds",
+                "Angle",
+                "TemperatureControllerPower",
+            )
+        )
+
+        # Listen to temperature monitor messages
+        pub.subscribe(
+            self.write, f"serial.{config.TEMPERATURE_MONITOR_TOPIC}.data.response"
+        )
+
+    def close(self) -> None:
+        """Close the current file handle."""
+        pub.unsubscribe(
+            self.write, f"serial.{config.TEMPERATURE_MONITOR_TOPIC}.data.response"
+        )
+        self._writer.close()
+
+    @pubsub_errors("data_file.error")
+    def write(self, time: datetime, temperatures: list[Decimal]) -> None:
+        """Write temperature readings to the CSV file."""
+        # Also include timestamp as seconds since midnight
+        midnight = datetime(time.year, time.month, time.day)
+        secs_since_midnight = floor((time - midnight).total_seconds())
+
+        self._writer.writerow(
+            (
+                time.strftime("%Y%m%d"),
+                time.strftime("%H:%M:%S"),
+                *temperatures,
```

Comment:
Do we want to add a format specifier here? Otherwise it will write the `Decimal` values with silly numbers of decimal places

Reply:
Good point. I'll look see what they did in the old program

Reply:
I've had a think about this and I think it's actually better if we *don't* use a format specifier. The real DP9800 has a fixed number of decimal places and we don't do any arithmetic with it so we should just end up with however many DPs the DP9800 gives us, which isn't many. Maybe let's just try it with the real device and see how the data look?

The downside is that we will end up with a silly no. of DPs for our simulated data.

Reply:
Ah yes, I guess because they're converted to `Decimal` from strings this should be fine